build.ps1

```diff
@@ -1,66 +1,140 @@
-                $exception = $_.Exception
-                throw $exception
-            }
-        }
-    }
-}
+<#
+.SYNOPSIS
+Build this repository
 
-cd $PSScriptRoot
+.DESCRIPTION
+Downloads korebuild if required. Then builds the repository.
 
-$repoFolder = $PSScriptRoot
-$env:REPO_FOLDER = $repoFolder
-$koreBuildZip="https://github.com/aspnet/KoreBuild/archive/dev.zip"
-if ($env:KOREBUILD_ZIP)
-{
-    $koreBuildZip=$env:KOREBUILD_ZIP
-}
+.PARAMETER Path
+The folder to build. Defaults to the folder containing this script.
+
+.PARAMETER Channel
+The channel of KoreBuild to download.
 
-$buildFolder = ".build"
-$buildFile="$buildFolder\KoreBuild.ps1"
+.PARAMETER DotNetHome
+The directory where .NET Core tools will be stored.
 
-if (!(Test-Path $buildFolder)) {
-    Write-Host "Downloading KoreBuild from $koreBuildZip"
+.PARAMETER ToolsSource
+The base url where build tools can be downloaded.
 
-    $tempFolder=$env:TEMP + "\KoreBuild-" + [guid]::NewGuid()
-    New-Item -Path "$tempFolder" -Type directory | Out-Null
+.PARAMETER Update
+Updates KoreBuild to the latest version even if a lock file is present.
 
-    $localZipFile="$tempFolder\korebuild.zip"
+.PARAMETER MSBuildArgs
+Arguments to be passed to MSBuild
 
-    DownloadWithRetry -url $koreBuildZip -downloadLocation $localZipFile -retries 6
+.NOTES
+This function will create a file $PSScriptRoot/korebuild-lock.txt. This lock file can be committed to source, but does not have to be.
+When the lockfile is not present, KoreBuild will create one using latest available version from $Channel.
 
-    Add-Type -AssemblyName System.IO.Compression.FileSystem
-    [System.IO.Compression.ZipFile]::ExtractToDirectory($localZipFile, $tempFolder)
+#>
+[CmdletBinding(PositionalBinding = $false)]
+param(
+    [Alias('p')]
+    [string]$Path = $PSScriptRoot,
+    [Alias('c')]
+    [string]$Channel = 'dev',
+    [Alias('d')]
+    [string]$DotNetHome = $(`
+            if ($env:DOTNET_HOME) { $env:DOTNET_HOME } `
+            elseif ($env:USERPROFILE) { Join-Path $env:USERPROFILE '.dotnet'} `
+            elseif ($env:HOME) {Join-Path $env:HOME '.dotnet'}`
+            else { Join-Path $PSScriptRoot '.dotnet'} ),
+    [Alias('s')]
+    [string]$ToolsSource = 'https://aspnetcore.blob.core.windows.net/buildtools',
+    [Alias('u')]
+    [switch]$Update,
+    [Parameter(ValueFromRemainingArguments = $true)]
+    [string[]]$MSBuildArgs
+)
 
-    New-Item -Path "$buildFolder" -Type directory | Out-Null
-    copy-item "$tempFolder\**\build\*" $buildFolder -Recurse
+Set-StrictMode -Version 2
+$ErrorActionPreference = 'Stop'
 
-    # Cleanup
-    if (Test-Path $tempFolder) {
-        Remove-Item -Recurse -Force $tempFolder
+#
+# Functions
+#
+
+function Get-KoreBuild {
+
+    $lockFile = Join-Path $Path 'korebuild-lock.txt'
+
+    if (!(Test-Path $lockFile) -or $Update) {
+        Get-RemoteFile "$ToolsSource/korebuild/channels/$Channel/latest.txt" $lockFile
     }
+
+    $version = Get-Content $lockFile -Tail 1
```

Comment:
I do think a version option would be a good idea. It seems trivial to add here but if it's more than just adding the switch and a couple lines, then it can wait.